### PR TITLE
Convert gem to Engine

### LINKED ISSUE #25: Convert the gem to a Rails Engine

### PR DESCRIPTION
**Pull request by jaybhoyar**
Closes #25

### CODE REVIEW

**Comment by jaybhoyar** [comment]
@chiraggshah _a Please review.

File: lib/neeto_notifications/notification.rb [hunk 1]
```diff
@@ -3,21 +3,38 @@
...
+        sanitized_responses = []
+        responses.each do |response|
+          sanitized_responses.push(
+            {
+              body: response.body,
+              status: response.status
+            }
+          )
+        end
+        sanitized_responses
```
Comment (chiraggshah) [inline lib/neeto_notifications/notification.rb:37]: Use `.map` instead

```ruby
responses.map do |response| 
  { body: response.body, status: response.status  } 
end
```

File: lib/neeto_notifications/notification.rb [hunk 2]
```diff
@@ -3,21 +3,38 @@
 module NeetoNotifications
   class Notification < Request
     def send(**attributes)
-      post("push_notifications", body: format(attributes))
+      responses = []
+      attributes[:user].devices.each do |device|
+        responses.push(post("push_notifications", body: format(attributes[:payload], device)))
+      end
```
Comment (chiraggshah) [inline lib/neeto_notifications/notification.rb:9]: Use `.map` instead

```ruby
responses = attributes[:user].devices.map do |device|
  post("push_notifications", body: format(attributes[:payload], device))
end
```

Comment (chiraggshah) [inline lib/neeto_notifications/notification.rb:9]: Use `.map` instead

```ruby
responses = attributes[:user].devices.map do |device|
  post("push_notifications", body: format(attributes[:payload], device))
end
```

Comment (chiraggshah) [reply lib/neeto_notifications/notification.rb:9]: You can sanitize it as well.

```ruby
attributes[:user].devices.map do |device|
  response = post("push_notifications", body: format(attributes[:payload], device))
  { body: response.body, status: response.status  } 
end
```

Then no need for the sanitize function

Comment (chiraggshah) [reply lib/neeto_notifications/notification.rb:9]: You can sanitize it as well.

```ruby
attributes[:user].devices.map do |device|
  response = post("push_notifications", body: format(attributes[:payload], device))
  { body: response.body, status: response.status  } 
end
```

Then no need for the sanitize function

File: lib/generators/neeto_notifications/install_generator.rb [hunk 3]
```diff
@@ -24,67 +24,21 @@ def copy_migration_file
...
-          copy_file "models/device.rb", Rails.root.join("app/models/device.rb")
-        end
-      end
-
       def add_device_user_association
         if File.exist?("app/models/organization_user.rb")
           inject_into_file(
             "app/models/organization_user.rb",
-            "  has_many :devices, dependent: :destroy\n",
+            "  has_many :devices, class_name: 'Device', dependent: :destroy\n",
```
Comment (chiraggshah) [inline lib/generators/neeto_notifications/install_generator.rb:31]: Is this change needed?

File: lib/generators/neeto_notifications/install_generator.rb [hunk 4]
```diff
@@ -24,67 +24,21 @@ def copy_migration_file
...
             before: "  has_many :permissions, through: :role"
           )
         elsif File.exist?("app/models/user.rb")
           inject_into_file(
             "app/models/user.rb",
-            "  has_many :devices, dependent: :destroy\n",
+            "  has_many :devices, class_name: 'Device', dependent: :destroy\n",
```
Comment (chiraggshah) [inline lib/generators/neeto_notifications/install_generator.rb:37]: Same here

File: app/controllers/api/v1/devices_controller.rb [hunk 5]
```diff
@@ -4,7 +4,7 @@ class Api::V1::DevicesController < Api::V1::BaseController
   before_action :load_device!, only: [:destroy]
 
   def create
-    device = current_user.devices.new(device_params)
+    device = current_user.devices.find_or_initialize_by(device_params)
```
Comment (chiraggshah) [inline app/controllers/api/v1/devices_controller.rb:7]: Now since we own devices_controller.rb and device.rb inside the engine, lets add tests for the controller and the model.

**Review by chiraggshah — COMMENTED** [review]
@jaybhoyar _a : Added comments.

**Comment by jaybhoyar** [comment]
@chiraggshah _a Please review.

**Review by chiraggshah — APPROVED** [review]
Looks good.
